### DISCUSSION TITLE
How to back up scripts with kit?

### DISCUSSION THREAD

**Discussion by jsjoeio** [johnlindquist/kit#70] [Q&A] [answered]
I know `~/.kenv` is a git repo. Let's say I wanted to back up my `/scripts` and/or `/lib` directory, what's the suggested approach for that?

**Comment by johnlindquist** [comment, marked as answer]
The app only runs the initial clone for the .kenv repo. After that, the
end-user can do whatever they like. (Unlike the .kit repo where the App
keeps it up-to-date by pulling tags)

I'm strongly considering using "degit" for .kenv so the user can git init
themselves. The user just needs to respect the directory conventions.

Advanced usage: .kit can generate multiple .kenv dirs anywhere on your
machine, although 99% of people will only use the App .kenv, it's totally
possible to manage many other headless .kenvs.

**Reply by jsjoeio** [reply]
Ah, gotcha!

> I'm strongly considering using "degit" for .kenv so the user can git init
themselves.

Yeah, this would help I think. I was confused because the `.kenv` is already git initialized so it threw me off and I wasn't sure if could point it to my own remote or something.
<img width="518" alt="image" src="https://user-images.githubusercontent.com/3806031/111919085-5944f500-8a45-11eb-9dd2-795a7ba0a4cf.png">

**Reply by jh3y** [reply]
I'm about to write up my experience with this 😃 

__TL;DR__ I set up a separate remote which worked great. But, there are a couple of quirks that I'm not sure about. But, I'm sure @johnlindquist will straighten me out with what I could be doing better 😄

**Reply by jsjoeio** [reply]
@jh3y Will you link the post back here when you do? 😄 It would definitely help to piggyback off your approach!

**Reply by jh3y** [reply]
Sure thing! I wrote about it in a separate thread: https://github.com/johnlindquist/kit/discussions/71

Some good actionable things from John to improve it. Overall, awesome experience and makes my machine set up a doddle. I'll refine it further, the scripts are very "rough". But, I was able to test it on three machines and it worked a treat once I worked out the steps I needed.

**Comment by lannonbr** [comment]
As I'm thinking it will be nice to store my scripts to share between my computers & have them publicly accessible, I'm saving my `~/.kenv/scripts/` folder up to git: [lannonbr/scriptkit-scripts](https://github.com/lannonbr/scriptkit-scripts).

The one thing I think I will need to easily sync things up between machines is make sure that all of the scripts have a related shell script within ~/.kenv/bin. I am digging through the code and would `~/.kit/cli/create-all-bins.js` help solve this after doing `git pull`?

**Reply by johnlindquist** [reply]
Yup, "generate bins" is an option in the help menu. It runs that 👍

**Reply by lannonbr** [reply]
As a followup, I did end up using this and it works wonderfully:

```js
// Menu: Sync Scripts
// Description: Pull from backup scripts git repo & create shell scripts for ones that don't have one yet
// Author: Benjamin Lannon
// Twitter: @lannonbr

exec('cd ~/.kenv/scripts && git pull origin main')
await cli('create-all-bins')
```

This doesn't resolve cleaning up `~/.kenv/bin` of scripts that no longer exist, but this is good enough to be able to quickly pull down scripts and have them ready to be used either in the terminal or the prompt

**Reply by cfritze64** [reply]
any advice on  the proper way to remove the ~/.kenv/lib directory from tracking so I can set up may own repo sync for ~/.kenv/lib  ?    Presently, If I got to the ~/.kenv/lib directory and "git remote -v" I see john's main scriptkit repo as origin. I'm loathe to mess with things, worried I'll upset the overall kit update structure.

**Reply by johnlindquist** [reply]
@cfritze64 An an early-adopter, you have a very old `.kenv`. 

The app now removes all `.git` info from both the `.kenv` and the `.kit`. You can safely delete `.kenv/.git`

`.kenv`: Change whatever you'd like, just not the directory structure
`.kit`: Don't touch anything, the app will completely wipe that dir with each update

**Reply by cfritze64** [reply]
That's great. I thought that might be the case. Thanks for the clear answer.
